server/libs/websocket.js

```diff
@@ -0,0 +1,72 @@
+const WebSocket = require('ws');
+const db = require('./database');
+const BSON = require('bson');
+const server = require('../bin/www').server;
+
+const bson = new BSON();
+const wss = new WebSocket.Server({server});
+
+const errors = {
+  CONNECT_ERROR: bson.serialize({Error: 'Node couldn\'t connect to PostgreSQL. See Node console for details.'}),
+  SELECT_ERROR: bson.serialize({Error: 'Node returned an error during the SELECT query. See Node console for details.'}),
+};
+
+wss.on('connection', function(ws, req) {
+  pool.connect(function(err, client, done) {
+    if (err) {
+      wss.send(errors.CONNECT_ERROR);
+      return console.error(db.errors.CONNECT_ERROR, err.stack);
+    }
+
+    // when client connects, select row with latest telemetry information
+    // as a JSON object and send it
+    const query = 'SELECT json_agg(latest) ' +
+                  'FROM (SELECT * ' +
+                        'FROM packet ' +
+                        'ORDER BY timestamp DESC LIMIT 1) latest';
+    client.query(query, function(err, result) {
+      done();
+        if (err) {
+          ws.send(errors.SELECT_ERROR);
+          return console.error(db.errors.SELECT_ERROR, err.stack);
+        }
+
+        // send latest telemetry row to client (array with 1 element)
+        wss.send(bson.serialize(result.rows));
+    });
+  });
+
+  // get_past_data event handler to view historical telemetry data
+  // MUST receive a JSON object with fields 'start' and 'end' with timestamps
+  ws.on('get_past_data', function(msg) {
```

Comment:
Im not 100% what the intended use for this for, but this shouldn't be the normal use case. The server should be keeping a list of all the clients and continuously sending the data to it. On connection, it should just be putting it in the connection in a list, and some other task is sending the current data to all clients every second or so.

Reply:
EDIT: Disregard, this comment is not relevant. Nguyen's design is good.